core/src/main/scala/kafka/controller/ControllerContext.scala

```diff
@@ -22,6 +22,45 @@ import org.apache.kafka.common.TopicPartition
 
 import scala.collection.{Map, Seq, Set, mutable}
 
+object PartitionReplicaAssignment {
+  def fromOldAndNewReplicas(oldReplicas: Seq[Int], newReplicas: Seq[Int]): PartitionReplicaAssignment = {
+    val fullReplicaSet = (newReplicas ++ oldReplicas).distinct
+    PartitionReplicaAssignment(
+      fullReplicaSet,
+      fullReplicaSet.filterNot(oldReplicas.contains(_)),
+      fullReplicaSet.filterNot(newReplicas.contains(_))
+    )
+  }
+}
+
+case class PartitionReplicaAssignment(replicas: Seq[Int], addingReplicas: Seq[Int], removingReplicas: Seq[Int]) {
+  def isBeingReassigned: Boolean = {
+    addingReplicas.nonEmpty || removingReplicas.nonEmpty
+  }
+
+  /**
+    * Returns the partition replica assignment previous to this one.
+    * It is different than this one only when the partition is undergoing reassignment
+    * Note that this will not preserve the original ordering
+    */
+  def previousAssignment: PartitionReplicaAssignment = {
+    PartitionReplicaAssignment(
+      replicas.filterNot(addingReplicas.contains(_)),
+      Seq(),
+      Seq()
+    )
+  }
+
+  /**
+    * Returns the target replica assignment for this partition.
+    * This is different than the `replicas` variable only when there is a reassignment going on
+    */
+  def targetReplicas: Seq[Int] = replicas.filterNot(removingReplicas.contains(_))
```

Comment:
Similarly to this one: would you please summarize why don't we take original + adding_replicas - removing_replicas as target? Does this have something to do with the2 phase algorithm that first we add AR and then remove RR? (And apologies if I missed something obvious.)

Reply:
You're misunderstanding the `replicas` set. It is not the original replicas. Let me give a quick example:

replicas = [1,2,3]
request with target_replicas = [4,5,6]
Upon processing the request, our PartitionReplicaAssignment class would look like this:
```
PartitionReplicaAssignment(
  replicas=[1,2,3,4,5,6]
  addingReplicas=[4,5,6]
  removingReplicas=[1,2,3]
)
```
You can see where we create the case class instance in `onPartitionReassignment`, specifically the `PartitionReplicaAssignment.fromOldAndNewReplicas` method which helps us create the new assignment.

Reply:
Thanks, this is helpful.